File: python/algo/202605/algo3801.py
from bisect import bisect_right
from cmath import inf
from functools import cache
from typing import List


class Solution:
    def minMergeCost(self, lists: List[List[int]]) -> int:
        n = len(lists)

        @cache
        def length(i: int) -> int:
            res = 0
            for j in range(n):
                if (i >> j) & 1: res += len(lists[j])
            return res

        @cache
        def median(i: int) -> int:
            cnt = (length(i) + 1) // 2

            l, r = inf, -inf
            for j in range(n):
                if (i >> j) & 1:
                    l = min(l, min(lists[j]))
                    r = max(r, max(lists[j]))
            l -= 1
            r += 1
            while l + 1 < r:
                res = 0
                mid = (l + r) // 2
                for j in range(n):
                    if (i >> j) & 1: res += bisect_right(lists[j], mid)
                if res >= cnt: r = mid
                else : l = mid
            med = -inf
            for j in range(n):
                if (i >> j) & 1: 
                    inx = bisect_right(lists[j], r) # r is result
                    if inx: med = max(med, lists[j][inx-1])
            return med  


        @cache
        def dfs(st: int) -> int:
            if st.bit_count() == 1: return 0
            
            i = (st - 1) & st
            ans = inf
            while i > st//2:
                ans = min(ans, dfs(i) + dfs(st - i) + length(st) + abs(median(st - i) - median(i)))
                i = (i-1) & st
            return ans

        return dfs((1 << n) -1)

File: cplus/leetcode/202605/algo3801.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    // translate from python by doubao
    long long minMergeCost(vector<vector<int>>& lists) {
        int n = lists.size();
        int max_mask = 1 << n;

        long long len[max_mask];
        int med[max_mask];
        long long dp[max_mask];
        
        memset(len, -1, sizeof(len));
        memset(med, -1, sizeof(med));
        memset(dp, -1, sizeof(dp));

        function<long long(int)> get_len = [&](int mask) -> long long {
            if (len[mask] != -1) return len[mask];
            long long res = 0;
            for (int j = 0; j < n; j++) {
                if (mask & (1 << j)) 
                    res += (long long)lists[j].size();
            }
            return len[mask] = res;
        };

        function<int(int)> get_med = [&](int mask) -> int {
            if (med[mask] != -1) return med[mask];
            long long cnt = (get_len(mask) + 1) / 2;

            int l = INT_MAX, r = INT_MIN;
            for (int j = 0; j < n; j++) {
                if (mask & (1 << j)) {
                    l = min(l, lists[j][0]);
                    r = max(r, lists[j].back());
                }
            }
            l--; r++;

            while (l + 1 < r) {
                long long mid = (1LL * l + r) / 2;
                long long res = 0;
                for (int j = 0; j < n; j++) {
                    if (mask & (1 << j)) {
                        res += upper_bound(lists[j].begin(), lists[j].end(), (int)mid) 
                               - lists[j].begin();
                    }
                }
                if (res >= cnt) r = (int)mid;
                else l = (int)mid;
            }

            int res = INT_MIN;
            for (int j = 0; j < n; j++) {
                if (mask & (1 << j)) {
                    int pos = upper_bound(lists[j].begin(), lists[j].end(), r) 
                              - lists[j].begin();
                    if (pos > 0) res = max(res, lists[j][pos - 1]);
                }
            }
            return med[mask] = res;
        };

        function<long long(int)> dfs = [&](int st) -> long long {
            if (__builtin_popcount(st) == 1) return 0;
            if (dp[st] != -1) return dp[st];

            long long ans = LLONG_MAX;
            for (int i = (st - 1) & st; i > st / 2; i = (i - 1) & st) {
                long long cost = dfs(i) + dfs(st - i) 
                               + get_len(st) 
                               + abs((long long)get_med(i) - get_med(st - i));
                ans = min(ans, cost);
            }
            return dp[st] = ans;
        };

        return dfs((1 << n) - 1);
    }
};